### ISSUE TITLE
Refactor components to use dateRange via datastore

### ISSUE THREAD

**Issue by aaemnnosttv** [google/site-kit-wp#1531]
## Feature Description

Once the datastore foundation is in place in #1529, the existing `DateRangeSelector` should be refactored to use the datastore for its state, rather than WP hooks

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* `DateRangeSelector` should use datastore state internally instead of the current filter-based implementation
* All other components/functions which consume/reference dateRange state (e.g. `dataAPI`) must be updated to source the value from the datastore, without using `@wordpress/hooks`

## Implementation Brief

Legacy non-functional components that can't be rewritten as functional components will need to be updated to be wrapped with [`withSelect`](https://github.com/WordPress/gutenberg/tree/master/packages/data#withSelect) and [`withDispatch`](https://github.com/WordPress/gutenberg/tree/master/packages/data#withDispatch) as needed which provide registry selected data or dispatch functions to the consuming components as props.  
This can be done by wrapping the component before it's exported.

### Update `DateRangeSelector`  
_should be reimplemented as functional component_

**Displaying all options**
Rather than defining all options internally, it can get them via the new `getAvailableDateRanges` utility

**Selecting a new date range**
Replace the current hook with a dispatch call when a new range is selected
https://github.com/google/site-kit-wp/blob/09a45d7c0f2a8af533335d8d7076e434f04051b5/assets/js/components/date-range-selector.js#L77-L84
* Replace with `useDispatch(...).setDateRange( slug )`

**Triggering new data to be fetched**
Currently this is done like so
https://github.com/google/site-kit-wp/blob/09a45d7c0f2a8af533335d8d7076e434f04051b5/assets/js/components/date-range-selector.js#L86-L88

This can be done with `useEffect( () => {...}, [ dateRange ] )` to run every time the date range changes

The component will still need to use WordPress hooks for this until we're no longer using the legacy `dataAPI`.

### `dataAPI` and utils

After all other components are refactored to use date range state from the datastore, `dataAPI` will be all that's left, using it via these utilities
* `getCurrentDateRange`
* `getCurrentDateRangeSlug`

These can be updated to use `slug = Data.select( 'core/user' ).getDateRange()` directly in place of the filters.
The date range label is available via `getAvailableDateRanges( slug ).label`

### Remaining components 
* `AnalyticsDashboardWidgetTopAcquisitionSources`  
	- clean refactor to functional component with `useSelect` as it only has a `render` method
* `AnalyticsDashboardWidget`
	- refactor with data HOCs to provide `dateRange`
* `GoogleSitekitSearchConsoleDashboardWidget`
	- refactor with data HOCs to provide `dateRange`

## QA Brief

* 

## Changelog entry

* Migrate the existing date range selector component to rely on the centrally managed date range from datastore.

**Comment by felixarntz** [comment]
@aaemnnosttv I wonder if we should create a new component and leave the existing one untouched. This ties into the old data API, and refactoring here could easily result in side effects unexpectedly breaking existing code. Plus refactoring will make this more complicated than it would otherwise be, and maybe we'll have to make weird trade-offs to cater for a legacy edge-case.

On the other hand, of course refactoring would result in bigger wins right away. For example, using React hooks in the legacy data API, how would that work? What if the value was `undefined`?

It'd be great to have a detailed IB here on refactoring the existing component and based on that we can assess the wins and risks and decide whether we should proceed or build from scratch.

**Comment by aaemnnosttv** [comment]
> For example, using React hooks in the legacy data API, how would that work?

We wouldn't use React hooks in the legacy `dataAPI` since it isn't technically a component.
We could call `Data.select()` directly if we needed to from a function such as `getCurrentDateRange`.  However, the `DateRangeSelector` component is the only place this filter is referenced outside of `assets/js/util/date-range.js`.

Aside from the legacy `dataAPI`, the only other places the `getCurrentDateRange*` functions are referenced are from components, which could easily be updated to use the state via the datastore.

> What if the value was undefined?

At this point the only remaining usage of these utilities would be in `dataAPI`. I think the easiest way to avoid this problem would be to prime the date range as part of the `dataAPI.init` (this method doesn't exist anymore but you get the idea).

I'll add an IB here for your consideration.

**Comment by felixarntz** [comment]
The IB mostly sounds good to me. I'd like to hold off approval until we've fully defined #1529 though.

**Comment by felixarntz** [comment]
@aaemnnosttv A few minor comments regarding IB, other than that ✅ 
* `getAllDateRanges` is now `getAvailableDateRanges`, and it's a utility function not in the datastore.
* `Data.select( 'core/user' ).getDateRange()` now returns `slug` instead of `{ slug, name }`, so code that needs the name must call the selector for `slug` and then rely on `getAvailableDateRanges` to look up `name`. We may want to introduce a helper `getAvailableDateRange( slug )` for that.
* Regarding the initial selector value of `undefined`, having another look, would this actually ever be the case? This selector is not sourced by an API call, so it shouldn't return `undefined` initially, it would immediately return `INITIAL_STATE.dateRange` right? If that is indeed a problem though, I'm okay just falling back to the default of "last-28-days". cc @tofumatt

**Comment by aaemnnosttv** [comment]
> Regarding the initial selector value of `undefined`, having another look, would this actually ever be the case? This selector is not sourced by an API call, so it shouldn't return `undefined` initially, it would immediately return `INITIAL_STATE.dateRange` right? If that is indeed a problem though, I'm okay just falling back to the default of "last-28-days".

Selectors are synchronous so they always return based on the current state. If the `INITIAL_STATE` was `last-28-days` it would return that on the first select even if it were sourced by an API call, but really any async resolution would result in a potentially `undefined` return, but only if the initial state was `undefined`. In the future, if we use local/session storage for persistence, these APIs are all async, so it could potentially return "stale" or out-of-sync state for that first request, although given how fast it would resolve, I'm not sure it would matter.

The AC for 1529 says "The initial `dateRange` should be `last-28-days` for now although in the future this may be persisted from a previous choice" so that sounds like initial state to me 😄

**Comment by felixarntz** [comment]
Right, but that means we don't need to handle this here, since `getDateRange` would never return `undefined` right?

Can you update the function names in the IB? After that, it's good from my end.

**Comment by aaemnnosttv** [comment]
> Can you update the function names in the IB? After that, it's good from my end.

@felixarntz updated! I removed the part about initial state as well since that will be handled in #1529 anyways.

Assuming this is IB ✅  - moving to backlog 👍

**Comment by aaemnnosttv** [comment]
CR ✅  - Assigning to @felixarntz to decide if we want to include it in the upcoming release or postpone the merge until after.

**Comment by tofumatt** [comment]
QA ✅  from me!
